design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,217 @@
+        "HelpText": "Enables or disables the BIOS configuration of Single Root
+                     I/O Virtualization (SR-IOV) devices. Enable this feature
+                     if booting to a virtualization operating system that
+                     recognize SR-IOV devices. ",
+        "Hidden": false,
+        "Immutable": false,
+        "MenuPath": "./IntegratedDevicesRef",
+        "ReadOnly": false,
+        "ResetRequired": true,
+        "Type": "Enumeration",
+        "Value": [
+          {
+            "ValueDisplayName": "Enabled",
+            "ValueName": "Enabled"
+          },
+          {
+            "ValueDisplayName": "Disabled",
+            "ValueName": "Disabled"
+          }
+        ],
+        "WarningText": null,
+        "WriteOnly": false
+    }
+
+The cached BIOS registry from Ironic should be made available to the user
+via an API. The BIOSRegistry can be quite large (up to 500 Kb uncompressed
+on some vendors) and is only useful on the reference host, not all of the hosts
+that the BIOS config will be pushed to.
+
+Using the BIOS settings and BIOS registry, the user will be able create a JSON
+list of bulk BIOS settings. An additional CRD will be defined describing these
+bulk settings. This JSON list can be set on multiple machines to load and
+apply, effectively cloning all the settings from another machine. The BMO will
+send the data to Ironic via the clean-steps API without validating or
+interpreting it. Ironic will convert the JSON to a set of clean steps and run
+cleaning to configure the BIOS settings.
+
+### Implementation Details/Notes/Constraints
+
+One of the primary drivers of this design was the desire to not have the
+BMO validate the user inputted bulk BIOS settings as it requires vendor
+specific checks that are difficult to maintain. This proposal does not suggest
+validity checks in the BMO or Ironic either - the input JSON list is converted
+to clean steps as-is.
+
+In order to avoid applying attributes that are ReadOnly when setting BIOS
+configuration, which will have no effect and will increase the time for BIOS
+configuration it is recommended that the user remove ReadOnly fields, as
+identified in the BIOS Registry, when building the JSON input. It’s not
+expected that the entire set read from Ironic will be passed down as:
+
+1. many of the values read from Ironic are ReadOnly
+2. setting ReadOnly values to their current settings increases the size of the
+   input and may potentially affect the Redfish request to the BMC
+
+The BIOS Settings are changed during the node cleaning step which requires a
+reboot. The change to BIOS settings will not be a trigger for reprovisioning a
+host, the BIOS settings will only be applied when a new image is going to be
+written. In order to prevent redundant BIOS settings to the BMC, the current
+BIOS settings should be compared to the bulk BIOS settings and the settings
+only written to Ironic when there is a difference.
```

Comment:
The BMO would need to do this filtering work?

Reply:
Yes, the current BIOS settings and new BIOS settings should be fairly small (10's of attributes), especially if ReadOnly fields are stripped out (compared to the BIOSRegistry which is quite large with descriptions, enum values etc.) per attribute. 

Or we could do something like every time the new BIOS settings are received send it to Ironic on the next provisioning.